File: src/test/java/com/github/bgora/CarDealerFacadeTest.java
package com.github.bgora;

import com.github.bgora.american.AmericanCarFactory;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

class CarDealerFacadeTest {

    private final CarDealerFacade tested;

    CarDealerFacadeTest() {
        tested = new CarDealerFacade();
    }

    @Test
    void shouldSellCar() {
        //given
        var givenCarName = "Ford";
        var giveCarHorse = 200;
        var givenCarMaxSpeed = 120;

        var americanCar = AmericanCarFactory.createCar(givenCarName, giveCarHorse, givenCarMaxSpeed);

        //when
        tested.importFromUSA(americanCar);
        var car = tested.sellCar(givenCarName);

        var expectedSpeed = 193;
        var exceptedString = "Car{" +
                "name='" + givenCarName + '\'' +
                ", horsePower=" + giveCarHorse +
                ", maxSpeed=" + expectedSpeed + "km/h"+
                '}';

        //then
        assertThat(car.toString()).isEqualTo(exceptedString);
        assertThat(car.maxSpeed()).isEqualTo(expectedSpeed);
    }

    @Test
    void shouldSellManyCars() {
        //given
        var givenCarNames = List.of("Ford", "Pontiac", "Chevrolet", "Dodge Viper");
        var givenCars = List.of("Ford;200;120", "Pontiac;360;130", "Chevrolet;100;100", "Dodge Viper;500;150");


        var americanCars = AmericanCarFactory.produce(givenCars);

        //when
        tested.importFromUSA(americanCars);
        var cars = tested.sellCars(givenCarNames);

        var expectedSpeed = List.of(193, 209, 160, 241);

        //then
        var index = 0;
        for (Car car : cars){
            assertThat(car.maxSpeed()).isEqualTo(expectedSpeed.get(index++));
        }
    }
}

File: src/test/kotlin/com/github/bgora/CarDealerFacadeTest.kt
package com.github.bgora

import com.github.bgora.american.AmericanCar
import com.github.bgora.american.AmericanCarFactory
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test

internal class CarDealerFacadeTest {

    private val tested: CarDealerFacade = CarDealerFacade()

    @Test
    fun shouldSellCar() {
        //given
        val givenCarName = "Ford"
        val giveCarHorse = 200
        val givenCarMaxSpeed = 120
        val americanCar: AmericanCar = AmericanCarFactory.createCar(givenCarName, giveCarHorse, givenCarMaxSpeed)

        //when
        tested.importFromUSA(americanCar)
        val car = tested.sellCar(givenCarName)
        val expectedSpeed = 193
        val exceptedString = "Car{" +
                "name='" + givenCarName + '\'' +
                ", horsePower=" + giveCarHorse +
                ", maxSpeed=" + expectedSpeed + "km/h" +
                '}'

        //then
        car.toString() shouldBe exceptedString
        car?.maxSpeed shouldBe  expectedSpeed

    }

    @Test
    fun shouldSellManyCars() {
        //given
        val givenCarNames: List<String> = listOf("Ford", "Pontiac", "Chevrolet", "Dodge Viper")
        val givenCars = listOf("Ford;200;120", "Pontiac;360;130", "Chevrolet;100;100", "Dodge Viper;500;150")
        val americanCars: List<AmericanCar> = AmericanCarFactory.produce(givenCars)

        //when
        tested.importFromUSA(americanCars)
        val cars = tested.sellCars(givenCarNames)
        val expectedSpeed = listOf(193, 209, 160, 241)

        //then
        var index = 0
        for ((_, _, maxSpeed) in cars!!) {
            maxSpeed shouldBe expectedSpeed[index++]
        }
    }
}